### PR TITLE
THREESCALE-10643 Opentelemetry support for 3scale operator

### PR DESCRIPTION
**Pull request by MStokluska**
Jira: https://issues.redhat.com/browse/THREESCALE-10643

This PR allows the opentelemetry to be enabled on managed apicasts.

**Verification**

**Installation**

- Create namespace

```
oc new-project threescale
```

- Install CRDs

```
make install
```

- Run the operator

```
make run
```

- Create new secret to be consumed by APIM

```
oc apply -f - <<EOF   
---
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: null
  name: s3-credentials
stringData:
  AWS_ACCESS_KEY_ID: something
  AWS_SECRET_ACCESS_KEY: something
  AWS_BUCKET: something
  AWS_REGION: us-east-1
type: Opaque
EOF
```

- Fetch cluster domain

```
DOMAIN=$(oc get routes console -n openshift-console -o json | jq -r '.status.ingress[0].routerCanonicalHostname' | sed 's/router-default.//')
```

- Create APIM

```
kubectl apply -f - <<EOF                                                        
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager-sample
  namespace: threescale
spec:
  system:
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: s3-credentials
  wildcardDomain: $DOMAIN
EOF
```

- Confirm 3scale is fully installed

```
oc wait --for=condition=available apimanager/apimanager-sample --timeout=-1s
```

**Prepare Jaeger instance**

- Create Jaeger instance:
```
make jaeger-deploy
```
- Port-forward Jaeger
```
kubectl port-forward service/jaeger 16686
```
- Open up Jaeger instance
```
navigate to http://127.0.0.1:16686
```
**Scenario 1 - Apicast Staging Opentelemetry**
- Create secret:
```
kubectl apply -f - <<EOF
---
apiVersion: v1
kind: Secret
metadata:
  name: otel-config
type: Opaque
stringData:
  config.json: |
    exporter = "otlp"
    processor = "simple"
    [exporters.otlp]
    # Alternatively the OTEL_EXPORTER_OTLP_ENDPOINT environment variable can also be used.
    host = "jaeger"
    port = 4317
    # Optional: enable SSL, for endpoints that support it
    # use_ssl = true
    # Optional: set a filesystem path to a pem file to be used for SSL encryption
    # (when use_ssl = true)
    # ssl_cert_path = "/path/to/cert.pem"
    [processors.batch]
    max_queue_size = 2048
    schedule_delay_millis = 5000
    max_export_batch_size = 512
    [service]
    name = "apicast-staging" # Opentelemetry resource name <in this example, apicast staging is used
EOF
```
- In apimanager, add the opentelemetry specification for staging apicast:
```
    stagingSpec:
      openTelemetry:
        enabled: true
        tracingConfigSecretRef:
          name: otel-config
``` 
- At this point, Apicast staging pod should be re-created with new volume, volume mount and 2 envs for opentelemtry.
- Navigate to 3scale portal, create sample product with backend (or use the default one coming from wizard)
- Ping the staging endpoint
- confirm that the response has header:
```
"Traceparent"
```
- Navigate to Jaeger, search for apicast-staging, your request should have been logged as "apicast-staging: apicast"

**Scenario 2 - Apicast Production Opentelemetry**
- Re-apply the secret, please note there's an additional key now
```
kubectl apply -f - <<EOF
---
apiVersion: v1
kind: Secret
metadata:
  name: otel-config
type: Opaque
stringData:
  config.json: |
    exporter = "otlp"
    processor = "simple"
    [exporters.otlp]
    # Alternatively the OTEL_EXPORTER_OTLP_ENDPOINT environment variable can also be used.
    host = "jaeger"
    port = 4317
    # Optional: enable SSL, for endpoints that support it
    # use_ssl = true
    # Optional: set a filesystem path to a pem file to be used for SSL encryption
    # (when use_ssl = true)
    # ssl_cert_path = "/path/to/cert.pem"
    [processors.batch]
    max_queue_size = 2048
    schedule_delay_millis = 5000
    max_export_batch_size = 512
    [service]
    name = "apicast-staging" # Opentelemetry resource name <in this example, apicast staging is used
  production.json: |
    exporter = "otlp"
    processor = "simple"
    [exporters.otlp]
    # Alternatively the OTEL_EXPORTER_OTLP_ENDPOINT environment variable can also be used.
    host = "jaeger"
    port = 4317
    # Optional: enable SSL, for endpoints that support it
    # use_ssl = true
    # Optional: set a filesystem path to a pem file to be used for SSL encryption
    # (when use_ssl = true)
    # ssl_cert_path = "/path/to/cert.pem"
    [processors.batch]
    max_queue_size = 2048
    schedule_delay_millis = 5000
    max_export_batch_size = 512
    [service]
    name = "apicast-production"
EOF
```
- add following configuration for apicast production in APIM
```
    productionSpec:
      openTelemetry:
        enabled: true
        tracingConfigSecretRef:
          name: otel-config
        tracingConfigSecretKey: "production.json"
```
- hit the production endpoint and confirm that it works the same way as in staging.

**Scenario 3**
- remove the Apicast staging configuration from apimanager
- confirm that otel volume and volume mounts have been removed, also confirm that the envs: OPENTELEMETRY and OPENTELEMETRY_CONFIG has been removed from apicast staging but not from production.

**Scenario 4**
- navigate to the otel secret and add label: `apimanager.apps.3scale.net/watched-by: "apimanager"`
- from the otel secret, remove the config.json key and value.
- confirm that the apicast production pod was re-deployed.

### CODE REVIEW

**Comment by austincunningham** [comment]
One small question in you verifcation steps you have secrets creating files like config.json

```
stringData:
  config.json: |
```
but the content after this point does't look like json format . Looks more like toml format, I am guessing the name of the file doesn't matter ?

File: config/jaeger/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: jaeger
```
Comment (austincunningham) [inline config/jaeger/deployment.yaml:5]: Just wondering if we need the Jaeger deployment , the docs seem to suggest its not supported , i.e. end users can install and configure along side of 3scale but we don't provide direct support for it https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.10/html/administering_the_api_gateway/introduction-to-advanced-operation-of-threescale-apicast-api-gateway_api-gateway-apicast#installing_jaeger_on_your_openshift_instance
having the deployment may confuse this point.

File: doc/gateway-instrumentation.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+Supported propagation types: [W3C](https://w3c.github.io/trace-context/)
+
+### Prerequisites
+
+* Opentelemetry Collector supporting the APIcast exporter. Currently, the only implemeneted [exporter](https://opentelemetry.io/docs/reference/specification/protocol/exporter/)
+in APIcast is OTLP over gRPC `OTLP/gRPC`. Even though OpenTelemetry specification supports also OTLP over HTTP (`OTLP/HTTP`),
+this exporter is not included in APIcast. It works with Jaeger since version **1.35**.
+
+For dev/testing purposes, you can deploy quick&easy Jaeger with few commands. **Not suitable** for production use, though.
```
Comment (austincunningham) [inline doc/gateway-instrumentation.md:17]: Guess this answers my question above ,  We need Jaeger, but we don't need the deployment and make commands. I'm in two minds about this one , doen't install stuff we don't directly support  vs dev experience and make it easy :thinking:

File: doc/gateway-instrumentation.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+```yaml
+kubectl apply -f - <<EOF
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: otel-config
+type: Opaque
+stringData:
+  config.json: |
```
Comment (austincunningham) [inline doc/gateway-instrumentation.md:46]: Made a comment about this in the verification steps about the content not being json.
I guess it's just a little confusing 
Easier to respond here

File: config/jaeger/deployment.yaml (hunk 1 above)
Comment (MStokluska) [reply config/jaeger/deployment.yaml:5]: This is only for the upstream operator. So that contributors can give it a try.

**Comment by austincunningham** [comment]
# Verification
## Scenario 1 - Apicast Staging Opentelemetry
- env vars added to new apicast staging pod
```bash
oc exec $(oc get po | grep Running | grep staging-2 | awk '{print $1}') -- env | grep OPEN
OPENTELEMETRY_CONFIG=/opt/app-root/src/otel-configs/config.json
OPENTELEMETRY=1
OPENSSL_VERIFY=false
```
- Traceparent is in the headers response `HTTP_TRACEPARENT`
```bash
curl "https://api-3scale-apicast-staging.apps.aucunnin.lpi0.s1.devshift.org:443/?user_key=4bc4fb0c07fa8778477ea919cef5d4ef"
{
  "method": "GET",
  "path": "/",
  "args": "user_key=4bc4fb0c07fa8778477ea919cef5d4ef",
  "body": "",
  "headers": {
    "HTTP_VERSION": "HTTP/1.1",
    "HTTP_HOST": "echo-api.3scale.net",
    "HTTP_TRACEPARENT": "00-087aa2396887ea60bdc89ae153afa444-5b1049bf410325bd-01",
    "HTTP_X_REAL_IP": "10.131.2.10",
    "HTTP_X_3SCALE_PROXY_SECRET_TOKEN": "Shared_secret_sent_from_proxy_to_API_backend_ef1de00cba99ede7",
    "HTTP_USER_AGENT": "curl/7.85.0",
    "HTTP_ACCEPT": "*/*",
    "HTTP_X_FORWARDED_HOST": "api-3scale-apicast-staging.apps.aucunnin.lpi0.s1.devshift.org",
    "HTTP_X_FORWARDED_PORT": "443",
    "HTTP_X_FORWARDED_PROTO": "https",
    "HTTP_FORWARDED": "for=185.167.180.230;host=api-3scale-apicast-staging.apps.aucunnin.lpi0.s1.devshift.org;proto=https",
    "HTTP_X_FORWARDED_FOR": "185.167.180.230,3.222.125.41",
    "HTTP_X_ENVOY_EXTERNAL_ADDRESS": "3.222.125.41",
    "HTTP_X_REQUEST_ID": "82ede367-ed09-40e1-9017-c4584dfdcdfe",
    "HTTP_X_ENVOY_EXPECTED_RQ_TIMEOUT_MS": "15000"
  },
  "uuid": "54fe4a40-a93d-4bb1-b359-80c78a2cef54"
}%                                                             
```
- apicast-staging is in jaeger
image

## Scenario 2 - Apicast Production Opentelemetry
- verify the env vars are applied to production
```bash
oc exec $(oc get po | grep Running | grep staging-2 | awk '{print $1}') -- env | grep OPEN
OPENTELEMETRY_CONFIG=/opt/app-root/src/otel-configs/config.json
OPENTELEMETRY=1
OPENSSL_VERIFY=false
```
- verify that Traceparent is in the headers response 
```bash
curl "https://api-3scale-apicast-production.apps.aucunnin.lpi0.s1.devshift.org/?user_key=4bc4fb0c07fa8778477ea919cef5d4ef" 
{
  "method": "GET",
  "path": "/",
  "args": "user_key=4bc4fb0c07fa8778477ea919cef5d4ef",
  "body": "",
  "headers": {
    "HTTP_VERSION": "HTTP/1.1",
    "HTTP_HOST": "echo-api.3scale.net",
    "HTTP_TRACEPARENT": "00-54bf749a2566deb1799c4d9fdacde7f6-fc1b72fac2217999-01",
    "HTTP_X_REAL_IP": "10.131.2.10",
    "HTTP_X_3SCALE_PROXY_SECRET_TOKEN": "Shared_secret_sent_from_proxy_to_API_backend_ef1de00cba99ede7",
    "HTTP_USER_AGENT": "curl/7.85.0",
    "HTTP_ACCEPT": "*/*",
    "HTTP_X_FORWARDED_HOST": "api-3scale-apicast-production.apps.aucunnin.lpi0.s1.devshift.org",
    "HTTP_X_FORWARDED_PORT": "443",
    "HTTP_X_FORWARDED_PROTO": "https",
    "HTTP_FORWARDED": "for=185.167.180.230;host=api-3scale-apicast-production.apps.aucunnin.lpi0.s1.devshift.org;proto=https",
    "HTTP_X_FORWARDED_FOR": "185.167.180.230,3.222.125.41",
    "HTTP_X_ENVOY_EXTERNAL_ADDRESS": "3.222.125.41",
    "HTTP_X_REQUEST_ID": "c74e85c9-8327-4222-b967-6f42df95b2b4",
    "HTTP_X_ENVOY_EXPECTED_RQ_TIMEOUT_MS": "15000"
  },
  "uuid": "c5023211-99ab-487b-9bb9-6863f033b037"
}%                                                               
```
- production appears in jaeger
image

## Scenario 3 remove staging
- volume mounts before
```bash
oc get po $(oc get po | grep Running | grep production-2 | awk '{print $1}') -o jsonpath='{range .spec.containers[*]}{.name}{"\n"}{range .volumeMounts[*]}{.name}{"\n"}{end}{end}' | sort -u 
apicast-production
kube-api-access-sqrkx
otel-volume
 oc get po $(oc get po | grep Running | grep staging-2 | awk '{print $1}') -o jsonpath='{range .spec.containers[*]}{.name}{"\n"}{range .volumeMounts[*]}{.name}{"\n"}{end}{end}' | sort -u
apicast-staging
kube-api-access-rdz2x
otel-volume
```
- env vars before
```bash
oc exec $(oc get po | grep Running | grep production-2 | awk '{print $1}') -- env | grep OPEN
Defaulted container "apicast-production" out of: apicast-production, system-master-svc (init)
OPENSSL_VERIFY=false
OPENTELEMETRY_CONFIG=/opt/app-root/src/otel-configs/production.json
OPENTELEMETRY=1
oc exec $(oc get po | grep Running | grep staging-2 | awk '{print $1}') -- env | grep OPEN
OPENTELEMETRY_CONFIG=/opt/app-root/src/otel-configs/config.json
OPENTELEMETRY=1
OPENSSL_VERIFY=false
```
- volume mounts after removing stagingSpec values
```bash
oc get po $(oc get po | grep Running | grep staging-3 | awk '{print $1}') -o jsonpath='{range .spec.containers[*]}{.name}{"\n"}{range .volumeMounts[*]}{.name}{"\n"}{end}{end}' | sort -u
apicast-staging
kube-api-access-kxfrj
 oc get po $(oc get po | grep Running | grep production-2 | awk '{print $1}') -o jsonpath='{range .spec.containers[*]}{.name}{"\n"}{range .volumeMounts[*]}{.name}{"\n"}{end}{end}' | sort -u
apicast-production
kube-api-access-sqrkx
otel-volume
```
- env var mounts after removign stagingSpec values
```bash
oc exec $(oc get po | grep Running | grep staging-3 | awk '{print $1}') -- env | grep OPEN
OPENSSL_VERIFY=false
➜  3scale-operator git:(THREESCALE-10643) ✗ oc exec $(oc get po | grep Running | grep production-2 | awk '{print $1}') -- env | grep OPEN
Defaulted container "apicast-production" out of: apicast-production, system-master-svc (init)
OPENSSL_VERIFY=false
OPENTELEMETRY_CONFIG=/opt/app-root/src/otel-configs/production.json
OPENTELEMETRY=1
```
- confirmed staging is removed from Jaeger UI
no its still present but this is to be expedected as the historical datapoints exist
image

## Scenario 4 add labels and remove config.json
- confimed after adding the label that a new deployment of apicast-production was created
```bash
oc get secret otel-config -o yaml | yq '.metadata.labels'
apimanager.apps.3scale.net/watched-by: apimanager
oc get po | grep Running | grep apicast-production-2
oc get po | grep Running | grep apicast-production-3
apicast-production-3-wxf9x    1/1     Running     0        
```
- from the otel-config secret, remove the config.json key and value and confirm another roll out of production
```bash
oc get secret otel-config -o yaml | yq '.data'       
production.json: ZXhwb3J0ZXIgPSAib3RscCIKcHJvY2Vzc29yID0gInNpbXBsZSIKW2V4cG9ydGVycy5vdGxwXQojIEFsdGVybmF0aXZlbHkgdGhlIE9URUxfRVhQT1JURVJfT1RMUF9FTkRQT0lOVCBlbnZpcm9ubWVudCB2YXJpYWJsZSBjYW4gYWxzbyBiZSB1c2VkLgpob3N0ID0gImphZWdlciIKcG9ydCA9IDQzMTcKIyBPcHRpb25hbDogZW5hYmxlIFNTTCwgZm9yIGVuZHBvaW50cyB0aGF0IHN1cHBvcnQgaXQKIyB1c2Vfc3NsID0gdHJ1ZQojIE9wdGlvbmFsOiBzZXQgYSBmaWxlc3lzdGVtIHBhdGggdG8gYSBwZW0gZmlsZSB0byBiZSB1c2VkIGZvciBTU0wgZW5jcnlwdGlvbgojICh3aGVuIHVzZV9zc2wgPSB0cnVlKQojIHNzbF9jZXJ0X3BhdGggPSAiL3BhdGgvdG8vY2VydC5wZW0iCltwcm9jZXNzb3JzLmJhdGNoXQptYXhfcXVldWVfc2l6ZSA9IDIwNDgKc2NoZWR1bGVfZGVsYXlfbWlsbGlzID0gNTAwMAptYXhfZXhwb3J0X2JhdGNoX3NpemUgPSA1MTIKW3NlcnZpY2VdCm5hbWUgPSAiYXBpY2FzdC1wcm9kdWN0aW9uIgo=
 oc get po | grep Running | grep apicast-production-3
oc get po | grep Running | grep apicast-production-4
apicast-production-4-68bzw    1/1     Running     0          62s
```

Verification works as expected , couple of open questions

File: config/jaeger/deployment.yaml (hunk 1 above)
Comment (austincunningham) [reply config/jaeger/deployment.yaml:5]: guess we can add the warning just like the docs team to the upstream docs 
>**WARNING:** Jaeger is a third-party component, which 3scale does not provide support for, with the exception of uses with APIcast. The following instructions are provided as a reference example only, and are not suitable for production use.

File: doc/gateway-instrumentation.md (hunk 3 above)
Comment (austincunningham) [reply doc/gateway-instrumentation.md:46]: suspect if it was just 
```yaml
stringData:
  config.toml: | 
```
it would work as well

Comment (MStokluska) [reply doc/gateway-instrumentation.md]: I've updated docs, thanks Austin

File: doc/gateway-instrumentation.md (hunk 2 above)
Comment (MStokluska) [reply doc/gateway-instrumentation.md:17]: I think we can follow for now what was done in apicast operator. The jaeger command won't work unless the upstream repo is used (for dev/testing only) so I think we are safe.

File: doc/gateway-instrumentation.md (hunk 3 above)
Comment (austincunningham) [reply doc/gateway-instrumentation.md]: yep confirmed with Michal that changing it to toml works.

File: pkg/3scale/amp/operator/apicast_options_provider.go [hunk 4]
```diff
@@ -532,3 +575,130 @@ func (a *ApicastOptionsProvider) setPodTemplateAnnotations() {
...
+	if !res.Enabled {
+		return res, nil
+	}
+
+	// In the APIcast CR validation step it is checked that when enabled, the secret ref is not nil
+	// Adding this to avoid panics
+	if a.apimanager.Spec.Apicast.StagingSpec.OpenTelemetry.TracingConfigSecretRef == nil || a.apimanager.Spec.Apicast.StagingSpec.OpenTelemetry.TracingConfigSecretRef.Name == "" {
+		fldPath := field.NewPath("spec").Child("apicast").Child("stagingSpec").Child("openTelemetry").Child("tracingConfigSecretRef")
+		errors := append(field.ErrorList{}, field.Invalid(fldPath, a.apimanager.Spec.Apicast.StagingSpec.OpenTelemetry.TracingConfigSecretRef, "tracing config secret name is empty"))
+		return res, errors.ToAggregate()
```
Comment (carlkyrillos) [inline pkg/3scale/amp/operator/apicast_options_provider.go:601]:
```suggestion
		errs := append(field.ErrorList{}, field.Invalid(fldPath, a.apimanager.Spec.Apicast.StagingSpec.OpenTelemetry.TracingConfigSecretRef, "tracing config secret name is empty"))
		return res, errs.ToAggregate()
```

There a few instances in this file where the variable name collides with the `errors` imported package name. It's probably a good idea to fix now if possible.

**Review by carlkyrillos — COMMENTED** [review]
Completed verification steps and review code changes. I left a few comments but feel free to ignore
/lgtm

**Pull request merged**